### DISCUSSION TITLE
Support specifying `onRecoverableError` for `ReactDOM.hydrateRoot`

### DISCUSSION THREAD

**Discussion by vpontis** [vercel/next.js#36641] [Ideas]
### Describe the feature you'd like to request

With the upgrade to React 18, we are getting a lot of hydration errors. These can be hard to debug on production because the source maps are minified. 

I'd like to see the ability to pass in `onRecoverableError` to [`hydrateRoot`](https://reactjs.org/docs/react-dom-client.html#hydrateroot) which will allow better logging of errors to help us track down hydration issues. (This would also let us ignore hydration errors if we decide to, or not send them to Sentry.)

### Describe the solution you'd like

You can define an `onRecoverableError` in `next.config.js` which will be called on [`hydrateRoot` ](https://github.com/vercel/next.js/search?q=hydrateRoot).

### Describe alternatives you've considered

I've tried tracking down hydration errors and I've found some of them, but not all of them.

**Comment by gurkerl83** [comment]
React handles recoverable errors in onRecoverableError. Either a custom function is passed to hydrateRoot; otherwise React's default implementation is used, which uses the browser's **reportError** function. **Errors that are recoverable in isolation are always reported to the user in a modal-style error dialogue, and are not handled quietly.**

**NextJs uses only two arguments** (Dom element, React element) when calling **hydrateRoot**. The **third** possible argument to specify **hydrate options** along with the **onRecoverableError** field, in which the error **response of recoverable errors** can be adjusted, **is not used at all**.

In principle, the problem can be addressed on the NextJs or React side, but an adjustment is necessary on the respective side.

**Option 1: NextJs - Extension of the NextJs configuration**
NextJs, as shown in the issue, does not allow an option in its configuration to specify hydrate options (specifically onRecoverableError) that are passed from NextJs to hydrateRoot.

**Con:** Valid only for NextJs, other libraries that hide the call to hydrateRoot from the user and do not allow hydrate options are still affected by the issue.

**Option 2: Adaptation of React's default implementation to onRecoverableError**

Errors related to suspense are called recoverable. These are resolved through React internal workflows by reverting to a client-based rendering within a Suspense Boundary. As these errors are handled, jumping back to a client-based rendering does not constitute a critical error. A hint in the console about the cause that led to the switch to a client-based rendering should be sufficient. 

```
Extract from ReactDOMRoot

const defaultOnRecoverableError =
   typeof reportError === 'function'
      ?  // In modern browsers, reportError will dispatch an error event, emulating an uncaught JavaScript error.
      reportError
      : (error: mixed) => {
         // In older browsers and test environments, fallback to console.error.
         // eslint-disable-next-line react-internal/no-production-logging, react-internal/warning-args.
      console.error(error);
      };
```

- Removal of browser-based API **reportError** to handle recoverable errors. 

https://developer.mozilla.org/en-US/docs/Web/API/reportError

As pointed out in the comment, an uncaught error is thrown even if it was classified as recoverable within React. React internal processes have responded to such errors, for example by server-side and client-side content not matching. 

The error caused by reportError must in turn be caught by a global error handler. This seems to be a bit excessive.

```
window.onerror = function(message, source, lineno, colno, error) {
  console.log('message:' + error.message + ', lineno: ' + lineno );
  return true;
};
```

- Output the error only in the console, as is the case in older browsers and test environments.

In React based libraries that do not allow or support configuration of hydrate options, the problem must be described as recurring. Especially in background that React notices recoverable errors and actively counteracts them, conversely should resort to a less aggressive behaviour to bring this to the developer's attention.

ReportError is a very young API - Supported from October 2021 in Edge, Firefox and Chrome, not before. 

https://caniuse.com/?search=reportError

LogRecoverableError was originally introduced in ReactDOMHostConfig. 
https://github.com/facebook/react/commit/848e802d203e531daf2b9b0edb281a1eb6c5415d

The following refactoring has changed the name and position.
https://github.com/facebook/react/commit/efd8f6442d1aa7c4566fe812cba03e7e83aaccc3

logRecoverableError => defaultOnRecoverableError
ReactDOMHostConfig => ReactDomRoot

I will open an issue in NextJs and React respectively.

https://github.com/facebook/react/issues/24563

Thx!

**Comment by BrooceLR** [comment]
This should really be converted back to an issue. Theres plenty of users confused by minified errors which can sometimes only be reproduced in production. Supporting `onRecoverableError` would make isolating, reporting and reproducing these issues *so much* easier.

**Comment by hamlim** [comment]
Hate to be the person to drop by and add a plus one to a discussion but we'd find this feature really valuable to help us debug these errors!

We'd be happy to also contribute this feature if given some guidance on the best place to support this (my _limited_ understanding is that next.config.js isn't loaded on the client, so we wouldn't be able to leverage that to define a function for this handler, I guess it could be defined in something like _app, or layout.tsx in the new infra [but that gets tricky if it's a server component 🤔 ])

**Comment by AbhiPrasad** [comment]
Would be great to see this get exposed. We've been chatting about it here https://github.com/facebook/react/issues/26224, and it seems Astro, Remix, Gatsby all already expose this functionality so Next.js just has to catch up!

if you want a super hacky approach, you can try and monkey patch `hydrateRoot` with proxies to accomplish this. https://github.com/getsentry/sentry-javascript/issues/7293. Warning though that this could break if next/react changes their API in any way.

```ts
// Run this code before react components get mounted.
// If you use Sentry you can put this in your sentry.client.config.js
import * as client from "react-dom/client";

const oldClientHydrateRoot = client.hydrateRoot;
client.hydrateRoot = new Proxy(oldClientHydrateRoot, {
  apply: (wrappingTarget, thisArg, args) => {
    const oldOnRecoverableError = args[2].onRecoverableError;

    args[2].onRecoverableError = new Proxy(oldOnRecoverableError, {
      apply: (inner_wrappingTarget, inner_thisArg, inner_args) => {
        const error = inner_args[0];
        const errorInfo = inner_args[1];
 
        // log out error + `componentStack` in production
        console.log(error, errorInfo.componentStack)

        // if y'all use Sentry (https://sentry.io/for/nextjs/)
        // Sentry.captureException(error, { contexts: { react: { componentStack } } });
    });

    return wrappingTarget.apply(thisArg, args);
  },
});
```

^ We use the above to get the component stack logged in production!

![image](https://user-images.githubusercontent.com/18689448/223114424-ff9e732a-d0fe-4ba5-b5ca-c0715e135cdc.png)

**Reply by JohnRPB** [reply]
I tried this and it didn't work for me, probably because I wasn't able to successfully monkey-patch the dom client before React mounted. But I found success instead monkey-patching the reportError() function in the client wrapper for my root layout:

```
const isBrowser = typeof window !== "undefined";

if (isBrowser) {
  // Obtain the `reportError` function from the global scope
  const reportError = window.reportError;
  // Monkey-patch the function to ignore specific errors
  window.reportError = (...args) => {
    const [ error ] = args;
    // Ignore error if it contains the text "strict-client-only"
    if (!error.message.includes("strict-client-only")) reportError(...args);
  }
}
```
This not only hides the console errors, it also prevents the popups, a little surprisingly. A different option for folks struggling with the first solution.

The error message BTW is unique to me, because I was throwing errors manually; you can adapt it to your use-case.

**Reply by AbhiPrasad** [reply]
So for anyone new who ran into this, you can add the monkeypatch I specified above in the new `instrumentation-client` Next.js API they just added: https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client

This should ensure that this runs before react gets mounted.

**Comment by AndersDJohnson** [comment]
Relates to https://github.com/vercel/next.js/discussions/47542.

**Comment by AndersDJohnson** [comment]
Principal Engineer on the Web team at @target here - we would find this extremely useful as well!

**Comment by AndersDJohnson** [comment]
Proposing https://github.com/vercel/next.js/pull/50943 as a potential partial solution here, at least for @AbhiPrasad's use case above, which my teams also have.

**Comment by iamkd** [comment]
I think it makes sense to have it exposed completely as well. There is a pattern [described in React docs](https://react.dev/reference/react/Suspense#providing-a-fallback-for-server-errors-and-client-only-content) that enables real client-only components without use of `dynamic` or `lazy`. The problem is that throwing an error will pollute the logs even though it is expected to be recoverable. 

Other frameworks like Remix do not hide the `hydrateRoot` call. This allows us to patch the `onRecoverableError` handler and ignore the expected recoverable errors or add better logging etc. Also confirmed by a member of React team in this issue: https://github.com/reactjs/react.dev/issues/6106#issuecomment-1674118195

**Reply by AbhiPrasad** [reply]
Let me shoot a ping to some Vercel folks about this, maybe we can revive some discussion and get some Next.js maintainers to leave their feedback.

**Reply by AbhiPrasad** [reply]
No dice - I think there's not much appetite to get this exposed.

fwiw I've been running the monkeypatch in https://github.com/vercel/next.js/discussions/36641#discussioncomment-5215494 for a while and it's been fine. Horrible hack, but better than nothing 🤷

**Reply by IGassmann** [reply]
It looks like the currently unstable [React's postpone API](https://github.com/facebook/react/blob/main/packages/react/src/ReactPostpone.js) is being created for this. See [example](https://gist.github.com/AndrewIngram/b7e2b6ddea0b0d55bc7ed44a227d9311)

**Reply by IGassmann** [reply]
It seems that it doesn't pollute the logs.

**Reply by karlhorky** [reply]
> It looks like the currently unstable [React's postpone API](https://github.com/facebook/react/blob/main/packages/react/src/ReactPostpone.js) is being created for this
>
> It seems that it doesn't pollute the logs.

@IGassmann thanks for this! However, I'm not following your comments above and how it relates to this discussion. You said the Postpone API is "created for this" - created for what? Can't see how this relates to better debugging and logging for hydration errors.

I think the majority of people contributing here are looking for ways to do better logging and debugging when hydration errors occur, as noted in this part of the original post at the top of the discussion:

> I'd like to see the ability to pass in onRecoverableError to [hydrateRoot](https://reactjs.org/docs/react-dom-client.html#hydrateroot) which will allow better logging of errors to help us track down hydration issues. (This would also let us ignore hydration errors if we decide to, or not send them to Sentry.)

Maybe I'm missing something? Would be great if you could expand on your comments above :)

**Reply by karlhorky** [reply]
Ah maybe you mean something like the description below from [the PR `facebook/react#27238`](https://github.com/facebook/react/pull/27238) (although I'm still not sure if this allows for handling hydration errors 🤔):

> Effectively `postpone(reason)` behaves like a built-in [Catch Boundary](https://github.com/facebook/react/pull/26854). It's like `raise(Postpone, reason)` except it's built-in so it needs to be able to be encoded and caught by Suspense boundaries.
> 
> In Flight and Fizz these behave pretty much the same as errors. Flight just forwards it to retrigger on the client. In Fizz they just trigger client rendering which itself might just postpone again or fill in the value. The difference is how they get logged.
> 
> In Flight and Fizz they log to `onPostpone(reason)` instead of `onError(error)`. This log is meant to help find deopts on the server like finding places where you fall back to client rendering. The reason that you pass in is for that purpose to help the reason for any deopts.
> 
> I do track the stack trace in DEV but I don't currently expose it to `onPostpone`. This seems like a limitation. It might be better to expose the Postpone object which is an Error object but that's more of an implementation detail. I could also pass it as a second argument.
> 
> On the client after hydration they don't get passed to `onRecoverableError`. There's no global `onPostpone` API to capture postponed things on the client just like there's no `onError`. At that point it's just assumed to be intentional. It doesn't have any `digest` or reason passed to the client since it's not logged.

**Reply by karlhorky** [reply]
> You said the Postpone API is "created for this"

Looking at @AndrewIngram's [original tweet](https://twitter.com/andrewingram/status/1718596092663603636) mentioning the gist, I guess it seems like what you meant by "this" was conditional rendering on client / server:

> ... to make a component “client only”. If you’re using the Next app router you can do one better and use React’s experimental postpone API

...which I don't think is the main point of this discussion, if I understand correctly.

**Reply by IGassmann** [reply]
> "this" was conditional rendering on client / server

That's it :), sorry for the confusion

> ...which I don't think is the main point of this discussion, if I understand correctly.

True. I saved this discussion some time ago to find a way to deal with the conditional rendering issues, without spamming the logs. I forgot this wasn't the main point of the discussion when I added that comment 😅

**Reply by JohnRPB** [reply]
The React recommendation in the OP is the exact thing I've been trying to implement, but using React.lazy. It works brilliantly well, and enables me to have components that do not revert to a suspended state at all if they have the right data and lazy-loaded code already in the browser, BUT I can't use Error Boundaries to catch the errors I'm throwing on the server, so in development I get numerous popups and on production there are technically errors thrown. Although invisible to the user, they hurt SEO on Google by damaging the Best Practices score.

If I used next/dynamic, I could get around this, but the dynamic loading api doesn't provide access to props, and my loading states sometimes depend on runtime information. So I'm going to have to implement something hacky like the hydrateRoot() suggestion above, or perhaps replacing window.reportError(). Thank god I found this discussion, which was only possible by searching Google for "onRecoverableError" after trawling through the next source code.

**Reply by JohnRPB** [reply]
Actually, since I'm using the experimental React release inside NextJS, the postpone API proved to be a much better approach and less of an anti-pattern:

```
import { unstable_postpone } from 'react';

function ClientOnlyComponent(props) {
    if (isServer) unstable_postpone("client only");
    ...
```

Thank you to the person who suggested this above, it worked wonders!

**Reply by JohnRPB** [reply]
Unfortunately I need to report that the above solution did not work, and only appeared to work because ironically it also threw an error.

In NextJS 14.0.4 (latest), unstable_postpone does not appear in the webpack-bundled import of React. This isn't documented anywhere, but I can see in the source code that NextJS is not exposing some experimental features out of what is, after all, still an unstable version of React, even though they are actually using them internally.

Rather than spend more time on this, I wanted to share the work-around I came up with, which involves monkey-patching the browser's reportError function:

```
const isBrowser = typeof window !== "undefined";

if (isBrowser) {
  // Obtain the `reportError` function from the global scope
  const reportError = window.reportError;
  // Monkey-patch the function to ignore specific errors
  window.reportError = (...args) => {
    const [ error ] = args;
    // Ignore error if it contains the text "strict-client-only"
    if (!error.message.includes("strict-client-only")) reportError(...args);
  }
}
```
This actually works better than I thought; it doesn't just hide the console errors, it prevents the popups too. I put it  in the client-wrapper for my root layout.

**Reply by AndrewIngram** [reply]
> Unfortunately I need to report that the above solution did not work, and only appeared to work because ironically it also threw an error.
> 
> 
> 
> In NextJS 14.0.4 (latest), unstable_postpone does not appear in the webpack-bundled import of React. This isn't documented anywhere, but I can see in the source code that NextJS is not exposing some experimental features out of what is, after all, still an unstable version of React, even though they are actually using them internally.
> 
> 
> 
> Rather than spend more time on this, I wanted to share the work-around I came up with, which involves monkey-patching the browser's reportError function:
> 
> 
> 
> ```
> 
> const isBrowser = typeof window !== "undefined";
> 
> 
> 
> if (isBrowser) {
> 
>   // Obtain the `reportError` function from the global scope
> 
>   const reportError = window.reportError;
> 
>   // Monkey-patch the function to ignore specific errors
> 
>   window.reportError = (...args) => {
> 
>     const [ error ] = args;
> 
>     // Ignore error if it contains the text "strict-client-only"
> 
>     if (!error.message.includes("strict-client-only")) reportError(...args);
> 
>   }
> 
> }
> 
> ```
> 
> This actually works better than I thought; it doesn't just hide the console errors, it prevents the popups too. I put it  in the client-wrapper for my root layout.

If you enable the ppr experiment, you also get access to the postpone API (because ppr uses it internally) - https://nextjs.org/docs/app/api-reference/next-config-js/partial-prerendering

**Reply by JohnRPB** [reply]
Thanks @AndrewIngram, I'll check this out! Might stick with the reportError() work-around for now given the bleeding-edge status of ppr, but maybe by the time I'm ready to push to production...

**Reply by JohnRPB** [reply]
Just wanted to throw in a note here that the Postpone API is tied to dynamic rendering, so keep that in mind; your pages will render dynamically with PPR if you enable it.

**Comment by samijaber** [comment]
Far from being an optimal solution, but I built a modal that shows a diff of client vs server rendered DOM when hydration errors occur: https://github.com/BuilderIO/hydration-overlay/

Somewhat of a workaround until this discussion bears fruit, or Hydration error handling improves within React itself.

**Reply by AbhiPrasad** [reply]
For production hydration errors you can use [Sentry Session Replay](https://sentry.io/for/session-replay/) to get a hydration error diff (both HTML and visual diff)! 

https://changelog.getsentry.com/announcements/diff-hydration-errors-with-replay

![image](https://github.com/vercel/next.js/assets/18689448/19629241-fdbc-4219-840f-e5e723074f0d)

**Comment by MikelArnaiz** [comment]
@AbhiPrasad I tried your hack and although I log some info I still unable to retrieve what caused the hydration error.

In Next 14 a diff was introduced that shows a diff with exactly what the server and the client rendered. It is also log to the console:

![Captura de pantalla 2024-09-30 a las 13 09 38](https://github.com/user-attachments/assets/09818384-3766-4f1b-8e8f-145cc9073c14)

But the info available `onRecoverableError` doesn't include the diff
![Captura de pantalla 2024-09-30 a las 13 33 26](https://github.com/user-attachments/assets/564e854a-2eb5-4aa0-93d1-dc9850a9eff9)


How do I get this diff? What else are you at doing to be able to display a diff in Sentry replay?

**Reply by AbhiPrasad** [reply]
> How do I get this diff? What else are you at doing to be able to display a diff in Sentry replay?

Unfortunately react doesn't expose this exact diff, and the error is a lot less useful in production environments as well.

I think Next.js does some extra logic to calculate this, but it's not publicly accessible API.

With sentry session replay it snapshots the HTML before and after the hydration error is thrown, which is how it can show the exact diff.

**Comment by shshaw** [comment]
My team and I spent days trying to track down production-only SSR hydration issues that would've been easily revealed with `onRecoverableError`, but Next prevented the information from being logged.

I finally found a trick from this post from @cmorten that lets you tap into `onRecoverableError` via Dev Tools overrides on the live site. https://medium.com/@craigmorten/how-to-debug-react-hydration-errors-5627f67a6548

Vercel, _please_ add a next.config.js option or some other API for a custom `onRecoverableError` callback.

**Reply by AbhiPrasad** [reply]
Given they just added support for a specific client-side instrumentation file, it seems like a good place to expose these new error hooks as well: https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client

I'll raise a draft PR for this next week, maybe with a sample PR functionality will be more likely to be added.

fwiw tanstack start exposes this OOO the box because they allow you to access [the react apis directly](https://tanstack.com/start/latest/docs/framework/react/learn-the-basics#the-client-entry-point). So does [remix/react-router](https://reactrouter.com/start/data/installation#create-a-router-and-render).

**Reply by ChellappanRajan** [reply]
> Given they just added support for a specific client-side instrumentation file, it seems like a good place to expose these new error hooks as well: https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client
> 
> 
> 
> I'll raise a draft PR for this next week, maybe with a sample PR functionality will be more likely to be added.
> 
> 
> 
> fwiw tanstack start exposes this OOO the box because they allow you to access [the react apis directly](https://tanstack.com/start/latest/docs/framework/react/learn-the-basics#the-client-entry-point). So does [remix/react-router](https://reactrouter.com/start/data/installation#create-a-router-and-render).

Hi @AbhiPrasad Using this solution, is it possible to compare the differences between the server and client to identify what is causing the hydration issue?

**Reply by AbhiPrasad** [reply]
@ChellappanRajan logging the error details from `onRecoverableError` will get you pretty far, but sometimes the referenced component or the attached `componentStack` can still be pretty confusing.

You'll also need some kind of sourcemapping to unminify the error and `componentStack`. If you're using Sentry, we do this automatically via a [`reactErrorHandler` API](https://docs.sentry.io/platforms/javascript/guides/react/usage/#capturing-errors) exposed from the Sentry React SDK. Note that again the below doesn't work for Next.js because `onRecoverableError` isn't exposed in anyway.

```js
import { createRoot } from "react-dom/client";

const container = document.getElementById(“app”);
const root = createRoot(container, {
  // Callback called when React automatically recovers from hydration errors
  onRecoverableError: Sentry.reactErrorHandler(),
});
root.render();
```

I did a talk about this at React Miami this year, but didn't see much appetite in getting this changed unfortunately :(
